File: app/src/main/java/com/zionhuang/music/extractor/JSFunction.java
package com.zionhuang.music.extractor;

import androidx.annotation.NonNull;

import com.google.gson.JsonElement;

public abstract class JSFunction extends JsonElement {
    public abstract JSON apply(JSON args) throws JSInterpreter.InterpretException;

    @Override
    public JsonElement deepCopy() {
        return null;
    }

    @NonNull
    @Override
    public String toString() {
        return "[Json Function]";
    }
}

File: app/src/main/java/com/zionhuang/music/youtube/utils/JsFunction.kt
package com.zionhuang.music.youtube.utils

import com.google.gson.JsonArray
import com.google.gson.JsonElement

class JsFunction(
        val block: (args: JsonArray) -> JsonElement
) : JsonElement() {
    operator fun invoke(args: JsonArray): JsonElement = block(args)
    override fun deepCopy(): JsonElement = this
    override fun toString(): String = "[JavaScript Function]"
}